### PR TITLE
Moving domain and python particles support

### LINKED ISSUE #433: Support for moving domains
Add support for moving domains: meshes whose coordinates change over time (e.g. WRF data sets such as Katrina). 

The lack of moving domain support poses problems in at least two areas:
1. When rendering a region-of-interest the camera does not move with the domain
2. The "Box" class used to select a region-of-interest does not maintain the same ROI relative to the origin of the mesh.

### PR DESCRIPTION
**Pull request by StasJ**
- Adds support for rendering particle datasets using the Python API
- Adds support for moving domains
    - Adds ability for camera to automatically track a moving domain
    - Adds ability for renderer extents to track a moving domain
    - These can be independently toggled under the viewpoint tab in the GUI and in the Camera class in the Python API
- Adds ability to query dataset timestep information using the Python API
- Fixes renderer domains unable to be expanded past current dataset bounds
- Fixes bug where datasets with time-varying dimensions would fail to load if they were opened using unprefixed filenames

Fix #433

### CODE REVIEW

**Comment by NihanthCW** [comment]
Note: Moving domain support is for all the renderers.

**Review by shaomeng — APPROVED** [review]
Regarding the moving domain component of this PR, it does move the domain. However, it fails to move the rendering along with the moving domain, as the screen recording shows. I think it's a necessary step to have automatic moving rendering so the software can be truly operational with this type of data sets. With that being said, I'm fine with either augment this PR, or open a new issue/PR to provide a moving rendering capability.


https://github.com/NCAR/VAPOR/assets/814885/a3ab0807-cd6f-499d-925b-f7df4e53bd90

**Comment by StasJ** [comment]
> Regarding the moving domain component of this PR, it does move the domain. However, it fails to move the rendering along with the moving domain, as the screen recording shows. I think it's a necessary step to have automatic moving rendering so the software can be truly operational with this type of data sets. With that being said, I'm fine with either augment this PR, or open a new issue/PR to provide a moving rendering capability.
> 
>  Screen.Recording.2023-08-10.at.12.32.29.PM.mov

You need to enable the moving domain components in the viewpoint tab.

s

File: lib/render/NavigationUtils.cpp
Comment (StasJ) [reply lib/render/NavigationUtils.cpp:282]: Timesteps are represented in as size_t and int interchangeably throughout Vapor. Your dataset would need to be on the order of hundreds to thousands of Petabytes before this becomes an issue.

Comment (shaomeng) [reply lib/render/NavigationUtils.cpp:282]: It's been used interchangeably is already a pretty bad sign. Any improvement needs to start from somewhere.

In any event, only providing a suggestion here.

**Comment by shaomeng** [comment]
> > Regarding the moving domain component of this PR, it does move the domain. However, it fails to move the rendering along with the moving domain, as the screen recording shows. I think it's a necessary step to have automatic moving rendering so the software can be truly operational with this type of data sets. With that being said, I'm fine with either augment this PR, or open a new issue/PR to provide a moving rendering capability.
> > Screen.Recording.2023-08-10.at.12.32.29.PM.mov
> 
> You need to enable the moving domain components in the viewpoint tab.
> s

This is pretty cool. Would it make sense to enable these two options automatically while rendering a moving domain?

File: lib/render/NavigationUtils.cpp
Comment (StasJ) [reply lib/render/NavigationUtils.cpp]: Fixed

**Comment by StasJ** [comment]
> > > Regarding the moving domain component of this PR, it does move the domain. However, it fails to move the rendering along with the moving domain, as the screen recording shows. I think it's a necessary step to have automatic moving rendering so the software can be truly operational with this type of data sets. With that being said, I'm fine with either augment this PR, or open a new issue/PR to provide a moving rendering capability.
> > > Screen.Recording.2023-08-10.at.12.32.29.PM.mov
> > 
> > 
> > You need to enable the moving domain components in the viewpoint tab.
> > s
> 
> This is pretty cool. Would it make sense to enable these two options automatically while rendering a moving domain?

I did not enable it automatically to both maintain the original behavior by default, and because it modifies the normal behavior of vapor: the data at each timestep is represented with the parameters established in the previous timestep. This new behavior breaks that norm since once it is enabled it changes the camera and renderer extents between time steps. I can also see the advantages of having it enabled by default, however, so I am fine enabling it if that is the consensus.

**Review by sgpearse — APPROVED** [review]
Thanks for fixing this ancient problem.  It works as I'd expect.

I think we should enable the tracking parameters by default because it wouldn't affect users without moving domains, and it would save the remaining users that do have moving domains some time.

I'll approve this and leave it up to you when you'd like to merge it.

**Comment by NihanthCW** [comment]
Thanks for the input guys.
I would recommend enabling the tracking params by default. As per Sam and Scott's point, I feel the current behavior is not what they would expect to see and could be a potential inconvenience.

**Comment by StasJ** [comment]
> Thanks for the input @sgpearse @StasJ @shaomeng .
> I would recommend enabling the tracking params by default. As per Sam and Scott's point, I feel the current behavior is not what they would expect to see and could be a potential inconvenience.

Fixed

**Review by NihanthCW — APPROVED** [review]

**Pull request merged**
